app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionController.scala

```diff
@@ -139,22 +150,32 @@ class SubscriptionController @Inject()(
       sessionMissingRedirect()
     })
 
-  def redirectSubscriptionResponse(either: Either[SubscriptionReturnedHttpError, (Arn, String)]): Result =
+  private def redirectSubscriptionResponse(either: Either[SubscriptionReturnedHttpError, (Arn, String)])(
+    implicit hc: HeaderCarrier): Future[Result] =
     either match {
       case Right((arn, _)) =>
         mark("Count-Subscription-Complete")
-        Redirect(routes.SubscriptionController.showSubscriptionComplete())
-          .flashing("arn" -> arn.value)
+        redirectUponSuccessfulSubscription(arn)
 
       case Left(SubscriptionReturnedHttpError(CONFLICT)) =>
         mark("Count-Subscription-AlreadySubscribed-APIResponse")
-        Redirect(routes.CheckAgencyController.showAlreadySubscribed())
+        Future.successful(Redirect(routes.CheckAgencyController.showAlreadySubscribed()))
 
       case Left(SubscriptionReturnedHttpError(status)) =>
         mark("Count-Subscription-Failed")
         throw new HttpException("Subscription failed", status)
     }
 
+  private[controllers] def redirectUponSuccessfulSubscription(arn: Arn)(implicit hc: HeaderCarrier) =
+    for (wasEligibileForMapping <- sessionStoreService.fetchMappingEligible) yield {
+      val redirectLocation = wasEligibileForMapping match {
+        case Some(true) => routes.SubscriptionController.showLinkAccount()
+        case _          => routes.SubscriptionController.showSubscriptionComplete()
+      }
+
+      Redirect(redirectLocation).withSession("arn" -> arn.value)
```

Comment:
This will remove the existing sessions. You may need to use `withSession(request + ("arn" -> arn.value)`

Reply:
Nice spot. Updated, thanks!